docs/asdf/developer_overview.rst

```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
+
+A note on tree traversal. One can traverse a tree in three ways:
+inorder, preorder, and postorder. These respectively mean whether
+nodes are visited in the horizontal ordering of the nodes displayed on 
+a graphs (inorder), descending the tree from the root, doing the left 
+node first, before the right node (preorder), or from the bottom up, doing
+both leaf nodes before the parent node (postorder). In generating the 
+pyyaml tree, preorder works since it builds the tree from the root
+as one would expect in constructing the tree. But in converting the 
+tagged tree into the custom tree, postorder is the natural course, where
+the children are generated first so that the parent node can refer to 
+the final objects.
+
+An important part of this conversion process is handled by an instance
+of the class ``treeutil._TreeModificationContext``. This class does much the
+same trick that pyyaml does with generators. Although pyyaml creates
+references between basic python objects, these references must be
+converted to references between ASDF objects, and doing so requires 
+a similar mechanism for building the ASDF objects. The 
+``_TreeModificationContext`` object (hereafter context object)
+holds the incomplete generators a similar way the pyyaml
+``construct_document`` function did. 
+
+There are differences though. It provides
+methods to indicate if nodes are pending (i.e., incomplete), and there
+is a special value ``PendingValue`` that is a signal that the node hasn't 
+been handled yet (e.g., it may be referencing something yet to be done).
+If ``PendingValue`` persists to the end, it indicates a failure to handle 
+circular references in the tag code. This approach was taken because 
+one of the earlier prototype implementations did something like this, 
+passing dict and list subclasses that would throw an exception if a
+``PendingValue`` element was accessed.  That would have been more friendly
+to extension developers, but it was discarded because it wasn't thougth
+it was worth turning all those high performance containers into slower
+asdf subclasses.  We may want to revisit this if we decide to implement
+a tree that tracks "dirty" nodes and only writes to disk those that
+have changed, since in that case we'll need custom container subclasses
+anyway.  We could also consider writing our own dict/list subclass in C
+so we could have our cake and eat it too.
+
+The ``walk_and_modify`` code handles the case where the tag code returns 
+a generator instead of a value. These generators are expected to be the
+similar kind of generator that pyyaml uses, but differing in that instead
+of returing an empty container object it will populate whatever elements
+it can completely (e.g, all non-mutable ones), and completing the 
+population of all the mutable memebers on the second iteration
+(which may, in turn, generate new generators for mutable elements 
+contained within). In handling these generators, the 
```

Comment:
Suggestion:
```
When it detects a generator, the ``walk_and_modify`` code retrieves
the first yielded value, then saves the generator in the context.
```

Reply:
Done, but you should check if the suggested text is saved in the right...context [sorry].